File: v1/src/main/java/com/api/v1/customer/utils/CustomerFinderUtilImpl.java
package com.api.v1.customer.utils;

import com.api.v1.customer.annotations.SSN;
import com.api.v1.customer.domain.Customer;
import com.api.v1.customer.domain.CustomerRepository;
import com.api.v1.customer.exceptions.CustomerNotFoundException;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;
import reactor.core.publisher.Mono;

@Component
class CustomerFinderUtilImpl implements CustomerFinderUtil {

    @Autowired
    private CustomerRepository repository;

    @Override
    public Mono<Customer> find(@SSN String ssn) {
        return repository
                .findAll()
                .filter(e -> e.getSsn().equals(ssn))
                .singleOrEmpty()
                .switchIfEmpty(Mono.error(new CustomerNotFoundException(ssn)));
    }

}

File: v2/src/main/kotlin/com/api/v2/CustomerFinderUtilImpl.kt
package com.api.v2

import kotlinx.coroutines.flow.firstOrNull
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Component

@Component
class CustomerFinderUtilImpl: CustomerFinderUtil {

    @Autowired
    private lateinit var customerRepository: CustomerRepository

    override suspend fun find(ssn: String): Customer {
        val customer = customerRepository
            .findAll()
            .firstOrNull()
        if (customer == null) throw CustomerNotFoundException()
        return customer
    }

}